### PR TITLE
Fix spurious image verification failure

### PR DESCRIPTION
**Pull request by wjt**
https://phabricator.endlessm.com/T13004

### CODE REVIEW

File: gnome-image-installer/pages/install/gis-install-page.c [hunk 1]
```diff
@@ -85,8 +85,8 @@ gis_install_page_prepare_read (GisPage *page, GError **error)
   basename = g_file_get_basename(priv->image);
   if (basename == NULL)
     {
-      *error = g_error_new(GIS_INSTALL_ERROR, 0, _("Image verification error: g_file_get_basename(\"%s\") failed"), priv->image);
-      return FALSE;
+      *error = g_error_new (GIS_INSTALL_ERROR, 0, _("Internal error"));
```
Comment (sgnn7) [inline gnome-image-installer/pages/install/gis-install-page.c]: Reducing the verbosity of the errors is probably going to cause much harder debugging in user-submitted error reports so I'd leave the originals as-is.

File: gnome-image-installer/pages/install/gis-install-page.c [hunk 2]
```diff
@@ -421,8 +429,55 @@ gis_install_page_verify (GisPage *page)
...
+
+static gboolean
+gis_install_page_prepare_verify (GisPage *page)
+{
+
+  GSubprocess *gpg_import;
+  GError *error = NULL;
+
+  gpg_import = g_subprocess_new (G_SUBPROCESS_FLAGS_NONE, &error,
+      "gpg", "--import", IMAGE_KEYRING, NULL);
```
Comment (sgnn7) [inline gnome-image-installer/pages/install/gis-install-page.c]: Seems pretty wrong that we import the whole keyring into the user's gpg keyring (even if this is the `root` user). If you want to go this route, use the `--homedir` param with a tempfile that you clean up afterwards

File: gnome-image-installer/pages/install/gis-install-page.c [hunk 3]
```diff
@@ -380,38 +388,38 @@ gis_install_page_gpg_progress (GIOChannel *source, GIOCondition cond, GisPage *p
...
-#define ARGS 12
-  gchar *args[ARGS] = { "gpg",
-                        "--enable-progress-filter", "--status-fd", "1",
-                        "--keyring", IMAGE_KEYRING,
-                        "--trusted-key", TRUSTED_KEYID,
-                        "--verify", "", "", NULL };
+  gchar *args[] = { "gpg",
+                    "--enable-progress-filter", "--status-fd", "1",
+                    "--trusted-key", TRUSTED_KEYID,
+                    "--verify", signature_path, image_path, NULL };
```
Comment (sgnn7) [inline gnome-image-installer/pages/install/gis-install-page.c]: See below but use a a separate keyring with `--homedir` if you go this route

File: gnome-image-installer/pages/install/gis-install-page.c (hunk 2 above)
Comment (sgnn7) [inline gnome-image-installer/pages/install/gis-install-page.c]: Actually, shouldn't the signing key already be somewhere on the computer? If so, importing is redundant.

File: gnome-image-installer/pages/install/gis-install-page.c [hunk 4]
```diff
@@ -421,8 +429,55 @@ gis_install_page_verify (GisPage *page)
...
+      gint code = g_subprocess_get_exit_status (gpg_import);
+      GError *error = g_error_new (GIS_INSTALL_ERROR, code,
+                                   _("Could not import Endless OS key"));
+      gis_install_page_verify_failed (page, error);
+    }
+
+  g_object_unref (gpg_import);
+}
+
+
```
Comment (sgnn7) [inline gnome-image-installer/pages/install/gis-install-page.c]: `dd`

**Comment by sgnn7** [comment]
@wjt Left a couple of comments

File: gnome-image-installer/pages/install/gis-install-page.c (hunk 1 above)
Comment (wjt) [inline gnome-image-installer/pages/install/gis-install-page.c]: These are only "original" as of 24 hours ago when I changed them from just "Image verification error." :smile: 

I'm in two minds on this: I don't want to add strings that are incomprehensible to translators; but you're right that it's easier to debug with more details. I guess I could use strings unmarked for translation for errors which should never occur?

File: gnome-image-installer/pages/install/gis-install-page.c (hunk 2 above)
Comment (wjt) [inline gnome-image-installer/pages/install/gis-install-page.c]: This is the `gnome-initial-setup` user, whose homedir is ephemeral; and this is part of the installer, not a persistent image.

The signing key is already on the system (at `IMAGE_KEYRING`) but if you just pass `--keyring` to GPG, it tries to write a temp file alongside that path, apparently…

If you know a way to do this in one `gpg` invocation… :)

File: gnome-image-installer/pages/install/gis-install-page.c (hunk 1 above)
Comment (sgnn7) [inline gnome-image-installer/pages/install/gis-install-page.c]: Yeah, translations complicate things but at the very least the long form of the error condition should go in some log output somewhere.

**Comment by wjt** [comment]
Updated:
- Single-shot call to `gpg` as discussed
- Verbose messages for programming errors, which will appear in the journal
- Much less code (as a result of the first bullet!)

**Comment by sgnn7** [comment]
@wjt LGTM

**Pull request merged**